ESM? Build errors after updating to v1.3.5

### What version of Remix are you using?

1.3.5 and 1.4.0

### Steps to Reproduce

update from 1.3.4 to 1.3.5 or 1.4.0

run `remix build`

### Expected Behavior

App should compile

### Actual Behavior

Lots of errors starting with the following:

```
Building Remix app in production mode...
y-indexeddb is possibly an ESM only package and should be bundled with "serverDependenciesToBundle in remix.config.js.
The path "encoding" is imported in generated/prisma/client/runtime/index.js but encoding is not listed in your package.json dependencies. Did you forget to install it?
✘ [ERROR] [plugin server-bare-modules] Cannot find module 'encoding/package.json'
Require stack:
- /node_modules/.pnpm/@remix-run+dev@1.3.5_25e4b5bc5190ce64f1c1869d744ce914/node_modules/@remix-run/dev/compiler/plugins/serverBareModulesPlugin.js
- /node_modules/.pnpm/@remix-run+dev@1.3.5_25e4b5bc5190ce64f1c1869d744ce914/node_modules/@remix-run/dev/compiler.js
- /node_modules/.pnpm/@remix-run+dev@1.3.5_25e4b5bc5190ce64f1c1869d744ce914/node_modules/@remix-run/dev/cli/commands.js
- /node_modules/.pnpm/@remix-run+dev@1.3.5_25e4b5bc5190ce64f1c1869d744ce914/node_modules/@remix-run/dev/cli/run.js
- /node_modules/.pnpm/@remix-run+dev@1.3.5_25e4b5bc5190ce64f1c1869d744ce914/node_modules/@remix-run/dev/cli/index.js
- /node_modules/.pnpm/@remix-run+dev@1.3.5_25e4b5bc5190ce64f1c1869d744ce914/node_modules/@remix-run/dev/index.js
- /node_modules/.pnpm/@remix-run+dev@1.3.5_25e4b5bc5190ce64f1c1869d744ce914/node_modules/@remix-run/dev/cli.js
    generated/prisma/client/runtime/index.js:15389:26:
      15389 │         convert = require("encoding").convert;
            ╵                           ~~~~~~~~~~
  This error came from the "onResolve" callback registered here:
    node_modules/.pnpm/@remix-run+dev@1.3.5_25e4b5bc5190ce64f1c1869d744ce914/node_modules/@remix-run/dev/compiler/plugins/serverBareModulesPlugin.js:46:12:
      46 │       build.onResolve({
         ╵             ~~~~~~~~~
```

There are many more errors, but all seem realeted.

Reverting back to v1.3.4 and everything is ok again.

Same here with fontawesome pro

```
[ERROR] [plugin server-bare-modules] Cannot find module '/Users/User/remix/node_modules/@fortawesome/pro-solid-svg-icons/package.json.js'
```

It adds `.js` when it should just look for `package.json`

Similar issue for me with importing Headless UI into a Remix Blues stack app. After upgrading to 1.3.5 from 1.3.4 I get the following types of errors:

```
$ npm run build

[... snip ...]

✘ [ERROR] [plugin server-bare-modules] Package subpath './package.json' is not defined by "exports" in /Users/zayenz/remix-app/node_modules/@headlessui/react/package.json

    node_modules/@remix-run/dev/compiler/plugins/serverBareModulesPlugin.js:125:32:
      125 │   let packageJsonFile = require.resolve(`${packageName}/package.j...
          ╵                                 ^

    at new NodeError (node:internal/errors:372:5)
    at throwExportsNotFound (node:internal/modules/esm/resolve:433:9)
    at packageExportsResolve (node:internal/modules/esm/resolve:712:3)
    at resolveExports (node:internal/modules/cjs/loader:482:36)
    at Function.Module._findPath (node:internal/modules/cjs/loader:522:31)
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:919:27)
    at Function.resolve (node:internal/modules/cjs/helpers:108:19)
    at warnOnceIfEsmOnlyPackage (/Users/zayenz/remix-app/node_modules/@remix-run/dev/compiler/plugins/serverBareModulesPlugin.js:125:33)
    at /Users/zayenz/remix-app/node_modules/@remix-run/dev/compiler/plugins/serverBareModulesPlugin.js:95:11
    at callback (/Users/zayenz/remix-app/@remix-run/dev/node_modules/esbuild/lib/main.js:920:34)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async handleRequest (/Users/zayenz/remix-app/node_modules/@remix-run/dev/node_modules/esbuild/lib/main.js:700:30)

  This error came from the "onResolve" callback registered here:

    node_modules/@remix-run/dev/compiler/plugins/serverBareModulesPlugin.js:46:12:
      46 │       build.onResolve({
         ╵             ~~~~~~~~~

    at setup (/Users/zayenz/remix-app/node_modules/@remix-run/dev/compiler/plugins/serverBareModulesPlugin.js:46:13)
    at handlePlugins (/Users/zayenz/remix-app/node_modules/@remix-run/dev/node_modules/esbuild/lib/main.js:842:23)

  The plugin "server-bare-modules" was triggered by this import

    route:/Users/zayenz/remix-app/app/routes/index.tsx:4:45:
      4 │ import { Disclosure, Menu, Transition } from "@headlessui/react";
        ╵                                              ~~~~~~~~~~~~~~~~~~~
```

Looks like `server-bare-modules` in `v0.0.0-experimental-9aba8b51` is failing to find `@heroicons/react`.

> ✘ [ERROR] [plugin server-bare-modules] Cannot find module '@heroicons/react'

```
  This error came from the "onResolve" callback registered here:

    node_modules/@remix-run/dev/compiler/plugins/serverBareModulesPlugin.js:47:12:
      47 │       build.onResolve({
         ╵             ~~~~~~~~~

    at setup (removed/node_modules/@remix-run/dev/compiler/plugins/serverBareModulesPlugin.js:47:13)
    at handlePlugins (removed/node_modules/esbuild/lib/main.js:842:23)
```

Here's a [repo](https://github.com/ianduvall/remix-test-1.3.5) to reproduce the issue (it's the Indie Stack with `@heroicons/react` added).

--- a/integration/compiler-test.ts
+++ b/integration/compiler-test.ts
@@ -1,6 +1,11 @@
 import { test, expect } from "@playwright/test";
 
-import { createFixture, createAppFixture, js } from "./helpers/create-fixture";
+import {
+  createFixture,
+  createAppFixture,
+  js,
+  json,
+} from "./helpers/create-fixture";
 import type { Fixture, AppFixture } from "./helpers/create-fixture";
 import { PlaywrightFixture } from "./helpers/playwright-fixture";
 
@@ -72,6 +77,13 @@ test.describe("compiler", () => {
             return <div id="esm-only-single-export">{esmOnlyPkg}</div>;
           }
         `,
+        "app/routes/package-with-submodule.jsx": js`
+          import { submodule } from "@org/package/sub-package";
+
+          export default function PackageWithSubModule() {
+            return <div id="package-with-submodule">{submodule()}</div>;
+          }
+        `,
         "remix.config.js": js`
           let { getDependenciesToBundle } = require("@remix-run/dev");
           module.exports = {
@@ -82,35 +94,74 @@ test.describe("compiler", () => {
             ],
           };
         `,
-        "node_modules/esm-only-pkg/package.json": `{
-          "name": "esm-only-pkg",
-          "version": "1.0.0",
-          "type": "module",
-          "main": "./esm-only-pkg.js"
-        }`,
+        "node_modules/esm-only-pkg/package.json": json`
+          {
+            "name": "esm-only-pkg",
+            "version": "1.0.0",
+            "type": "module",
+            "main": "./esm-only-pkg.js"
+          }
+        `,
         "node_modules/esm-only-pkg/esm-only-pkg.js": js`
           export default "esm-only-pkg";
         `,
-        "node_modules/esm-only-exports-pkg/package.json": `{
-          "name": "esm-only-exports-pkg",
-          "version": "1.0.0",
-          "type": "module",
-          "exports": {
-            ".": "./esm-only-exports-pkg.js"
+        "node_modules/esm-only-exports-pkg/package.json": json`
+          {
+            "name": "esm-only-exports-pkg",
+            "version": "1.0.0",
+            "type": "module",
+            "exports": {
+              ".": "./esm-only-exports-pkg.js"
+            }
           }
-        }`,
+        `,
         "node_modules/esm-only-exports-pkg/esm-only-exports-pkg.js": js`
           export default "esm-only-exports-pkg";
         `,
-
-        "node_modules/esm-only-single-export/package.json": `{
-          "name": "esm-only-exports-pkg",
-          "version": "1.0.0",
-          "type": "module",
-          "exports": "./esm-only-single-export.js"
-        }`,
+        "node_modules/esm-only-single-export/package.json": json`
+          {
+            "name": "esm-only-exports-pkg",
+            "version": "1.0.0",
+            "type": "module",
+            "exports": "./esm-only-single-export.js"
+          }
+        `,
         "node_modules/esm-only-single-export/esm-only-single-export.js": js`
           export default "esm-only-single-export";
+        `,
+        "node_modules/@org/package/package.json": json`
+          {
+            "name": "@org/package",
+            "version": "1.0.0"
+          }
+        `,
+        "node_modules/@org/package/sub-package/package.json": json`
+          {
+            "module": "./esm/index.js",
+            "sideEffects": false
+          }
+        `,
+        "node_modules/@org/package/sub-package/esm/package.json": json`
+          {
+            "type": "module",
+            "sideEffects": false
+          }
+        `,
+        "node_modules/@org/package/sub-package/esm/index.js": js`
+          export { default as submodule } from "./submodule.js";
+        `,
+        "node_modules/@org/package/sub-package/esm/submodule.js": js`
+          export default function submodule() {
+            return "package-with-submodule";
+          }
+        `,
+        "node_modules/@org/package/sub-package/index.js": js`
+          module.exports.submodule = require("./submodule.js");
+        `,
+        "node_modules/@org/package/sub-package/submodule.js": js`
+          module.exports = function submodule() {
+            return "package-with-submodule";
+          }
         `,
       },
     });
@@ -209,6 +260,16 @@ test.describe("compiler", () => {
     // rendered the page instead of the error boundary
     expect(await app.getHtml("#esm-only-exports-pkg")).toBe(
       '<div id="esm-only-exports-pkg">esm-only-exports-pkg</div>'
+    );
+  });
+
+  test("allows consumption of packages with sub modules", async ({ page }) => {
+    let app = new PlaywrightFixture(appFixture, page);
+    let res = await app.goto("/package-with-submodule", true);
+    expect(res.status()).toBe(200); // server rendered fine
+    // rendered the page instead of the error boundary
+    expect(await app.getHtml("#package-with-submodule")).toBe(
+      '<div id="package-with-submodule">package-with-submodule</div>'
     );
   });
 });

--- a/packages/remix-dev/compiler/plugins/serverBareModulesPlugin.ts
+++ b/packages/remix-dev/compiler/plugins/serverBareModulesPlugin.ts
@@ -68,7 +68,6 @@ export function serverBareModulesPlugin(
           onWarning &&
           !isNodeBuiltIn(packageName) &&
           !/\bnode_modules\b/.test(importer) &&
-          !builtinModules.includes(packageName) &&
           !dependencies[packageName]
         ) {
           onWarning(
@@ -135,7 +134,7 @@ function warnOnceIfEsmOnlyPackage(
   fullImportPath: string,
   onWarning: (msg: string, key: string) => void
 ) {
-  let packageDir = resolveModuleBasePath(packageName);
+  let packageDir = resolveModuleBasePath(packageName, fullImportPath);
   let packageJsonFile = path.join(packageDir, "package.json");
 
   if (!fs.existsSync(packageJsonFile)) {
@@ -171,8 +170,9 @@ function warnOnceIfEsmOnlyPackage(
 }
 
 // https://github.com/nodejs/node/issues/33460#issuecomment-919184789
-function resolveModuleBasePath(packageName: string) {
-  let moduleMainFilePath = require.resolve(packageName);
+// adapted to use the fullImportPath to resolve sub packages like @heroicons/react/solid
+function resolveModuleBasePath(packageName: string, fullImportPath: string) {
+  let moduleMainFilePath = require.resolve(fullImportPath);
 
   let packageNameParts = packageName.split("/");
 

